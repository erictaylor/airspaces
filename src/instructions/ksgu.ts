import { buildAirspace } from '../build-airspace';
import { SGU_AIRPORT_COORDINATES } from '../constants';
import {
  addDegreesToBearing,
  getBearing,
  getLatLonPoint,
  lawOfCosines,
  nauticalMilesToKilometers,
  pythagoreanTheorem,
  solveForAngleA,
  subtractDegreesFromBearing,
} from '../utils';

type ExtensionArgs = [bearing: number, distance: number, side: number];

export const RADIUS = 4.5;
export const NE_EXTENSION: ExtensionArgs = [30, 7.7, 1];
export const SW_EXTENSION: ExtensionArgs = [203, 8.5, 2];
const CEILING = 17999;
const FLOOR = 'SFC';

export const getInstruction = (radius: number): readonly string[] => {
  return buildAirspace(
    {
      airspaceClass: 'E',
      name: 'Saint George Class E2',
      ceiling: CEILING,
      floor: FLOOR,
    },
    [
      ['V', { variable: 'X', value: SGU_AIRPORT_COORDINATES }],
      ['DC', radius],
    ],
  );
};

export const getExtensionInstructions = (
  bearing: number,
  distance: number,
  side: number,
): readonly string[] => {
  const pointY = getLatLonPoint(
    SGU_AIRPORT_COORDINATES,
    bearing,
    nauticalMilesToKilometers(distance),
  );

  const pointA = getLatLonPoint(
    pointY,
    subtractDegreesFromBearing(bearing, 90),
    nauticalMilesToKilometers(side),
  );
  const pointB = getLatLonPoint(
    pointY,
    addDegreesToBearing(bearing, 90),
    nauticalMilesToKilometers(side),
  );

  const distanceFromAirportToBPoint = pythagoreanTheorem(side, distance);
  const angleA =
    90 - solveForAngleA(distance, side, distanceFromAirportToBPoint);
  const distanceFromBPointToCPoint = lawOfCosines(
    RADIUS,
    distanceFromAirportToBPoint,
    angleA,
  );

  const pointC = getLatLonPoint(
    pointA,
    addDegreesToBearing(bearing, 180),
    nauticalMilesToKilometers(distanceFromBPointToCPoint),
  );
  const pointD = getLatLonPoint(
    pointB,
    addDegreesToBearing(bearing, 180),
    nauticalMilesToKilometers(distanceFromBPointToCPoint),
  );

  const bearingFromAirportToPointC = getBearing(
    SGU_AIRPORT_COORDINATES,
    pointC,
  );
  const bearingFromAirportToPointD = getBearing(
    SGU_AIRPORT_COORDINATES,
    pointD,
  );

  return buildAirspace(
    {
      airspaceClass: 'E',
      name: `Saint George E4 ${bearing > 0 && bearing < 90 ? 'NE' : 'SW'}`,
      ceiling: CEILING,
      floor: FLOOR,
    },
    [
      ['DP', pointB],
      ['DP', pointA],
      ['V', { variable: 'X', value: SGU_AIRPORT_COORDINATES }],
      [
        'DA',
        {
          radius: RADIUS,
          angleStart: bearingFromAirportToPointC,
          angleEnd: bearingFromAirportToPointD,
        },
      ],
    ],
  );
};

export const main = () => {
  console.log('* St. George Regional Airport Class E2 Airspace');
  console.log(getInstruction(RADIUS).join('\n'));
  console.log('\n');
  // NE Extension
  console.log(
    '* St. George Regional Airport Class E4 Airspace - Northeast Segment',
  );
  console.log(getExtensionInstructions(...NE_EXTENSION).join('\n'));
  console.log('\n');
  // SW Extension
  console.log(
    '* St. George Regional Airport Class E4 Airspace - Southwest Segment',
  );
  console.log(getExtensionInstructions(...SW_EXTENSION).join('\n'));
};
