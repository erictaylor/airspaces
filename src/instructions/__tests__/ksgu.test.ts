import { describe, expect, it } from 'bun:test';
import { NE_EXTENSION, RADIUS, SW_EXTENSION, getExtensionInstructions, getInstruction } from '../ksgu';

describe('KSGU Instructions', () => {
  it('should return the expected primary instructions', () => {
    expect(getInstruction(RADIUS)).toMatchSnapshot();
  });

  it('should return the expected northeast extension instructions', () => {
    expect(getExtensionInstructions(...NE_EXTENSION)).toMatchSnapshot();
  });

  it('should return the expected southwest extension instructions', () => {
    expect(getExtensionInstructions(...SW_EXTENSION)).toMatchSnapshot();
  });
});
