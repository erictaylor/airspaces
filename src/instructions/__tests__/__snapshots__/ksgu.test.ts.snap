// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`KSGU Instructions should return the expected primary instructions 1`] = `
[
  "AC E",
  "AN SAINT GEORGE CLASS E2",
  "AH 17999 ft",
  "AL SFC",
  "V X=037:02:11.00 N 113:30:37.00 W",
  "DC 4.5",
]
`;

exports[`KSGU Instructions should return the expected northeast extension instructions 1`] = `
[
  "AC E",
  "AN SAINT GEORGE E4 NE",
  "AH 17999 ft",
  "AL SFC",
  "DP 037:08:20.34 N 113:24:42.65 W",
  "DP 037:09:20.23 N 113:26:52.79 W",
  "V X=037:02:11.00 N 113:30:37.00 W",
  "DA 4.5, 17.071185, 42.995148",
]
`;

exports[`KSGU Instructions should return the expected southwest extension instructions 1`] = `
[
  "AC E",
  "AN SAINT GEORGE E4 SW",
  "AH 17999 ft",
  "AL SFC",
  "DP 036:55:09.09 N 113:37:03.68 W",
  "DP 036:53:35.49 N 113:32:27.89 W",
  "V X=037:02:11.00 N 113:30:37.00 W",
  "DA 4.5, 176.585560, 229.357024",
]
`;
